### PR TITLE
[automodel] fix validation_step

### PR DESCRIPTION
**Pull request by soluwalana**
# What does this PR do ?

The default `hf_auto_model_for_causal_lm` recipe fails to run due to the model being loaded on CPU, the validation loss being logged twice with different values, the validation_loss not being returned from the hf_auto_model_for_causal_lm validation step, and the validation_step passing the wrong parameters to forward.

**Collection**: llm

# Changelog

- Removes the hardcoded "cpu" device for loading the model
- Fixes the validation loss in hf_auto_model_for_causal to ensure that it calls forward with the correct parameter
- Returns loss from valdiation_step in hf_auto_model_for_causal
- Adds a check on loss in FSDP2 to ensure that None losses are not logged.

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py [hunk 1]
```diff
@@ -176,7 +176,7 @@ def _configure_model(self, attn_implementation):
             return auto_cls.from_pretrained(
                 self.model_name,
                 torch_dtype=torch.bfloat16,
-                device_map="cpu",
+                device_map=self.device,
```
Comment (akoumpa) [inline nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py:179]: Hi, I want to keep the "cpu" value since setting this to self.device may result in OOM for models that don't fit completely on the GPU.

File: nemo/lightning/pytorch/strategies/fsdp2_strategy.py [hunk 2]
```diff
@@ -434,15 +435,8 @@ def lightning_module_state_dict(self) -> Dict[str, Any]:
         Only returns a non-empty state dict on rank 0 if ``save_distributed_checkpoint=False``.
 
         """
-        from nemo.lightning.pytorch.strategies.utils import to_cpu
-
         assert self.lightning_module is not None
         state_dict = self.lightning_module.state_dict()
-
-        module_names = list(state_dict.keys())
```
Comment (akoumpa) [inline nemo/lightning/pytorch/strategies/fsdp2_strategy.py:442]: I wonder what's the motivation for this case? Do you recall if is there a value which break `to_cpu` or if it was something else?

File: nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py (hunk 1 above)
Comment (soluwalana) [reply nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py:179]: This seems pretty unreasonable because it isn't overridable and is the default way this model gets created when using the recipe.

File: nemo/lightning/pytorch/strategies/fsdp2_strategy.py (hunk 2 above)
Comment (soluwalana) [reply nemo/lightning/pytorch/strategies/fsdp2_strategy.py:442]: I have not experienced any issues with this *not* being on CPU.  I think that this was set this way due to the fact that the device map was hardcoded to "CPU"

**Comment by akoumpa** [comment]
I will split this PR into two, one to fix the validation_step and another to optimize the lightning_module_state_dict function. I want to make sure lightning_module_state_dict does not break anything before proceeding.

**Review by akoumpa — APPROVED** [review]
LGTM, thanks.
